@madhusudancs, pre-submit pr tests seems to be failing.
```
I0309 17:30:55.995] Extracting kubernetes-test.tar.gz into /workspace/kubernetes
W0309 17:31:02.791] 2017/03/09 17:31:02 util.go:110: Step './get-kube.sh' finished in 15.948444991s
W0309 17:31:02.791] 2017/03/09 17:31:02 util.go:108: Running: ./federation/cluster/federation-down.sh
W0309 17:31:02.828] Project: k8s-jkns-pr-bldr-e2e-gce-fdrtn
W0309 17:31:02.829] Zone: us-central1-f
I0309 17:31:02.993] +++ [0309 17:31:02] Unjoining cluster "federation-e2e-gce-us-central1-b" from federation "e2e-federation"
**W0309 17:31:03.100] error: context "e2e-federation" does not exist**
W0309 17:31:03.104] 2017/03/09 17:31:03 util.go:110: Step './federation/cluster/federation-down.sh' finished in 312.820638ms
W0309 17:31:03.134] 2017/03/09 17:31:03 main.go:153: Saved XML output to /workspace/_artifacts/junit_runner.xml.
W0309 17:31:03.135] 2017/03/09 17:31:03 main.go:187: Something went wrong: error tearing down previous federation control plane: exit status 1
W0309 17:31:03.136] + chmod -R o+r /workspace/_artifacts
I0309 17:31:04.039] Starting pull-kubernetes-federation-e2e-gce-59...
W0309 17:31:04.071] Traceback (most recent call last):
W0309 17:31:04.071]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 409, in <module>
W0309 17:31:04.071]     main(create_parser().parse_args())
W0309 17:31:04.071]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 346, in main
W0309 17:31:04.072]     mode.start(runner_args)
W0309 17:31:04.072]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 239, in start
W0309 17:31:04.072]     check(*cmd)
W0309 17:31:04.072]   File "/var/lib/jenkins/workspace/pull-kubernetes-federation-e2e-gce/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 39, in check
W0309 17:31:04.072]     subprocess.check_call(cmd)
W0309 17:31:04.072]   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
W0309 17:31:04.072]     raise CalledProcessError(retcode, cmd)
```
Should it be fixed in this pr?

--- a/federation/cluster/common.sh
+++ b/federation/cluster/common.sh
@@ -416,4 +416,10 @@ function cleanup-federation-api-objects {
   $host_kubectl delete pods,svc,rc,deployment,secret -lapp=federated-cluster
   # Delete all resources in FEDERATION_NAMESPACE.
   $host_kubectl delete pvc,pv,pods,svc,rc,deployment,secret --namespace=${FEDERATION_NAMESPACE} --all
+  $host_kubectl delete ns ${FEDERATION_NAMESPACE}
+
+  # Poll until the namespace is completely gone.
+  while $host_kubectl get namespace ${FEDERATION_NAMESPACE} >/dev/null 2>&1; do
+    sleep 5
+  done
 }

--- a/federation/cluster/federation-down.sh
+++ b/federation/cluster/federation-down.sh
@@ -24,9 +24,27 @@ KUBE_ROOT=$(readlink -m $(dirname "${BASH_SOURCE}")/../../)
 # $KUBEDNS_CONFIGMAP_NAME and $KUBEDNS_CONFIGMAP_NAMESPACE.
 source "${KUBE_ROOT}/federation/cluster/common.sh"
 
+# federation_clusters returns a list of all the clusters in
+# federation, if at all the federation control plane exists
+# and there are any clusters registerd.
+function federation_clusters() {
+  if clusters=$("${KUBE_ROOT}/cluster/kubectl.sh" \
+      --context="${FEDERATION_KUBE_CONTEXT}" \
+      -o jsonpath --template '{.items[*].metadata.name}' \
+      get clusters) ; then
+    echo ${clusters}
+    return
+  fi
+  echo ""
+}
+
 # unjoin_clusters unjoins all the clusters from federation.
 function unjoin_clusters() {
-  for context in $(federation_cluster_contexts); do
+  # Unjoin only those clusters that are registered with the
+  # given federation. This is slightly different than
+  # joining clusters where we join all the clusters in the
+  # current kubeconfig with the "federation" prefix.
+  for context in $(federation_clusters); do
     kube::log::status "Unjoining cluster \"${context}\" from federation \"${FEDERATION_NAME}\""
 
     "${KUBE_ROOT}/federation/develop/kubefed.sh" unjoin \
@@ -48,6 +66,7 @@ if cleanup-federation-api-objects; then
   # cloud provider cleanups are implemented in the individual test
   # `AfterEach` blocks.
   # Also, we wait only if the cleanup succeeds.
+  kube::log::status "Waiting for 2 minutes to allow controllers to clean up federation components..."
   sleep 2m
 else
   echo "Couldn't cleanup federation api objects"

--- a/federation/cluster/federation-up.sh
+++ b/federation/cluster/federation-up.sh
@@ -79,6 +79,9 @@ function init() {
   local -r kube_registry="${KUBE_REGISTRY:-gcr.io/${project}}"
   local -r kube_version="$(get_version)"
 
+  kube::log::status "DNS_ZONE_NAME: \"${DNS_ZONE_NAME}\", DNS_PROVIDER: \"${DNS_PROVIDER}\""
+  kube::log::status "Image: \"${kube_registry}/hyperkube-amd64:${kube_version}\""
+
   "${KUBE_ROOT}/federation/develop/kubefed.sh" init \
       "${FEDERATION_NAME}" \
       --host-cluster-context="${HOST_CLUSTER_CONTEXT}" \

--- a/hack/federated-ginkgo-e2e.sh
+++ b/hack/federated-ginkgo-e2e.sh
@@ -22,13 +22,9 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
 
 source "${KUBE_ROOT}/cluster/kube-util.sh"
 
-#A little hack to get the last zone. we always deploy federated cluster to the last zone.
-#TODO(colhom): deploy federated control plane to multiple underlying clusters in robust way
-lastZone=""
-for zone in ${E2E_ZONES};do
-    lastZone="$zone"
-done
+: "${FEDERATION_HOST_CLUSTER_ZONE?Must set FEDERATION_HOST_CLUSTER_ZONE env var}"
+
 (
-    set-federation-zone-vars "$zone"
+    set-federation-zone-vars "${FEDERATION_HOST_CLUSTER_ZONE}"
     "${KUBE_ROOT}/hack/ginkgo-e2e.sh" $@
 )

